### PR TITLE
Use defaulted constructors and destructors

### PR DESCRIPTION
**Pull request by masterleinad**
This should look a bit cleaner (and generate less assembly).

### CODE REVIEW

File: core/unit_test/TestViewAPI.hpp [hunk 1]
```diff
@@ -919,7 +919,7 @@ class TestViewAPI {
       dView4_unmanaged;
   typedef typename dView0::host_mirror_space host;
 
-  TestViewAPI() {}
+  TestViewAPI() = default;
```
Comment (dalg24) [inline core/unit_test/TestViewAPI.hpp]: Remove.

File: core/unit_test/TestMemoryPool.hpp [hunk 2]
```diff
@@ -474,7 +474,7 @@ void test_memory_pool_corners(const bool print_statistics,
 
 template <class DeviceType, class Enable = void>
 struct TestMemoryPoolHuge {
-  TestMemoryPoolHuge() {}
+  TestMemoryPoolHuge() = default;
```
Comment (dalg24) [inline core/unit_test/TestMemoryPool.hpp]: Remove.

File: algorithms/src/Kokkos_Sort.hpp [hunk 3]
```diff
@@ -445,7 +445,7 @@ struct BinOp3D {
   typename KeyViewType::non_const_value_type range_[3];
   typename KeyViewType::non_const_value_type min_[3];
 
-  BinOp3D() {}
+  BinOp3D() = default;
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp]: Remove.

File: algorithms/src/Kokkos_Sort.hpp [hunk 4]
```diff
@@ -200,7 +200,7 @@ class BinSort {
   bool sort_within_bins;
 
  public:
-  BinSort() {}
+  BinSort() = default;
```
Comment (dalg24) [inline algorithms/src/Kokkos_Sort.hpp]: Remove.

File: core/src/Threads/Kokkos_ThreadsExec.cpp [hunk 5]
```diff
@@ -77,7 +77,7 @@ void (*volatile s_current_function)(ThreadsExec &, const void *);
 const void *volatile s_current_function_arg = nullptr;
 
 struct Sentinel {
-  Sentinel() {}
+  Sentinel() = default;
```
Comment (dalg24) [inline core/src/Threads/Kokkos_ThreadsExec.cpp]: Unnecessary too.

File: core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp [hunk 6]
```diff
@@ -211,7 +211,7 @@ inline T atomic_compare_exchange(
   union U {
     long i;
     T t;
-    KOKKOS_INLINE_FUNCTION U(){};
+    KOKKOS_INLINE_FUNCTION U() {}
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Compare_Exchange_Strong.hpp]: Unrelated changes (that were already fixed in #2717)

File: core/src/impl/Kokkos_Atomic_Exchange.hpp [hunk 7]
```diff
@@ -294,7 +294,7 @@ inline void atomic_assign(volatile T* const dest,
   union U {
     T val_T;
     type val_type;
-    inline U(){};
+    inline U() {}
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Exchange.hpp]: Done in #2717

File: core/src/impl/Kokkos_Atomic_Fetch_Add.hpp [hunk 8]
```diff
@@ -267,7 +267,7 @@ inline T atomic_fetch_add(volatile T* const dest,
   union U {
     long i;
     T t;
-    inline U(){};
+    inline U() {}
```
Comment (dalg24) [inline core/src/impl/Kokkos_Atomic_Fetch_Add.hpp]: Done in #2717

File: core/src/Kokkos_Cuda.hpp [hunk 9]
```diff
@@ -198,8 +198,8 @@ class Cuda {
   //--------------------------------------------------
   //! \name  Cuda space instances
 
-  KOKKOS_INLINE_FUNCTION
-  ~Cuda() {}
+  KOKKOS_DEFAULTED_FUNCTION
+  ~Cuda() = default;
```
Comment (dalg24) [inline core/src/Kokkos_Cuda.hpp]: Why would that destructor need to be `__host__ __device__` in the  first place?

File: core/src/OpenMP/Kokkos_OpenMP_Exec.hpp [hunk 10]
```diff
@@ -128,7 +128,7 @@ class OpenMPExec {
 
 namespace Kokkos {
 
-inline OpenMP::OpenMP() noexcept {}
+inline OpenMP::OpenMP() noexcept = default;
```
Comment (dalg24) [inline core/src/OpenMP/Kokkos_OpenMP_Exec.hpp]: Drop and remove declaration in `core/src/Kokkos_OpenMP.hpp`

**Review by dalg24 — CHANGES_REQUESTED** [review]

File: containers/unit_tests/TestDynViewAPI.hpp [hunk 11]
```diff
@@ -705,7 +705,7 @@ class TestDynViewAPI {
 
   typedef typename View0::host_mirror_space host_view_space;
 
-  TestDynViewAPI() {}
+  TestDynViewAPI() = default;
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply containers/unit_tests/TestDynViewAPI.hpp]: Yes, I was thinking of going through all the `defaulted` members in a second step.

File: algorithms/src/Kokkos_Sort.hpp (hunk 4 above)
Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp]: There is another constructor.

File: algorithms/src/Kokkos_Sort.hpp (hunk 3 above)
Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp]: There is another constructor.

File: algorithms/src/Kokkos_Sort.hpp (hunk 4 above)
Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp]: So?  It's in `Imp::` and as far as I know we don't need it.  (Correct me if I am wrong)

Comment (masterleinad) [reply algorithms/src/Kokkos_Sort.hpp]: No, it's not in namespace `Impl`. 
The constructor was introduced in 62e4ec0179dd205b0f1553398bf35395b68f00d1 but I don't know the reason. I couldn't find any use of it on `GitHub` but would not want to break backward compatibility in this pull request.

Comment (dalg24) [reply algorithms/src/Kokkos_Sort.hpp]: Fair enough.

@crtrott what was the rational for adding these?

Comment (crtrott) [reply algorithms/src/Kokkos_Sort.hpp]: I don't remember ...

Comment (crtrott) [reply algorithms/src/Kokkos_Sort.hpp]: Ok actually I think this was necessary to make some stuff working in LAMMPS I believe, where we actually may store a Sorter inside another class. And that class has a default constructor.

**Review by dalg24 — APPROVED** [review]

**Comment by dhollman** [comment]
It looks like we need to get on the same page about design guidelines for default constructors in Kokkos. I recommend reading [C++ core guideline C.43](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines.html#c43-ensure-that-a-copyable-value-type-class-has-a-default-constructor) on this. In particular, I'd like to draw attention to the following quote:

> The absence of a default value can cause surprises for users and complicate its use, so if one can be reasonably defined, it should be.

Based on this advice, I don't know that we need to go around removing default constructors from Kokkos, at least for now. We should have good reasons for not defining default constructors (though I can see how this advice might apply a little less strongly in HPC, since default initialization is often expensive for large value-like types, but particularly when we're dealing with reference semantics, I think it's fine to leave the default constructors alone).

**Comment by dalg24** [comment]
> I don't know that we need to go around removing default constructors from Kokkos

Look more carefully at the changes and you'll see that what the right thing to do.

File: core/src/Threads/Kokkos_ThreadsExec.cpp [hunk 12]
```diff
@@ -77,8 +77,6 @@ void (*volatile s_current_function)(ThreadsExec &, const void *);
 const void *volatile s_current_function_arg = nullptr;
 
 struct Sentinel {
-  Sentinel() {}
```
Comment (dalg24) [inline core/src/Threads/Kokkos_ThreadsExec.cpp]: This change is responsible for #2775

**Comment by dhollman** [comment]
> Look more carefully at the changes and you'll see that what the right thing to do.

Yeah, my comment was mostly in response to https://github.com/kokkos/kokkos/pull/2715#discussion_r378304826
